semtypes/src/main/java/io/ballerina/types/subtypedata/XmlSubtype.java

```diff
@@ -68,62 +73,46 @@ public static XmlSubtype from(int primitives, Bdd sequence) {
-            ro = (ro instanceof AllOrNothingSubtype) ? ro : makeSequence(true, (XmlSubtype) ro);
-
-            SubtypeData rw = Core.getComplexSubtypeData(cct, UniformTypeCode.UT_XML_RW);
-            rw = (rw instanceof AllOrNothingSubtype) ? rw : makeSequence(false, (XmlSubtype) rw);
-
-            return createXmlSemtype(ro, rw);
+            SubtypeData xmlSubtype = Core.getComplexSubtypeData(cct, BasicTypeCode.BT_XML);
+            xmlSubtype = (xmlSubtype instanceof AllOrNothingSubtype) ?
+                    xmlSubtype : makeXmlSequence((XmlSubtype) xmlSubtype);
+            return createXmlSemtype(xmlSubtype);
         }
     }
 
-    private static SubtypeData makeSequence(boolean roPart, XmlSubtype d) {
+    private static SubtypeData makeXmlSequence(XmlSubtype d) {
         int primitives = XML_PRIMITIVE_NEVER | d.primitives;
-        int atom = d.primitives &
-                (roPart ? XML_PRIMITIVE_RO_SINGLETON : XML_PRIMITIVE_SINGLETON);
+        int atom = d.primitives & XML_PRIMITIVE_SINGLETON;
         Bdd sequence = BddCommonOps.bddUnion(BddCommonOps.bddAtom(RecAtom.createRecAtom(atom)), d.sequence);
-        return createXmlSubtype(roPart, primitives, sequence);
+        return createXmlSubtype(primitives, sequence);
     }
 
-    public static ComplexSemType createXmlSemtype(SubtypeData ro, SubtypeData rw) {
-        ArrayList<UniformSubtype> subtypes = new ArrayList<>();
-        int all = 0;
-        if (ro instanceof AllOrNothingSubtype) {
-            if (Common.isAllSubtype(ro)) {
-                all = 1 << UniformTypeCode.UT_XML_RO.code;
-            }
-        } else {
-            subtypes.add(UniformSubtype.from(UniformTypeCode.UT_XML_RO, (XmlSubtype) ro));
-        }
-        if (rw instanceof AllOrNothingSubtype) {
-            if (Common.isAllSubtype(rw)) {
-                all |= 1 << UniformTypeCode.UT_XML_RO.code;
-            }
+    public static SemType createXmlSemtype(SubtypeData xmlSubtype) {
+        if (xmlSubtype instanceof AllOrNothingSubtype allOrNothingSubtype) {
+            return allOrNothingSubtype.isAllSubtype() ? PredefinedType.XML : PredefinedType.NEVER;
         } else {
-            subtypes.add(UniformSubtype.from(UniformTypeCode.UT_XML_RW, (XmlSubtype) rw));
+            return PredefinedType.basicSubtype(BasicTypeCode.BT_XML, (ProperSubtypeData) xmlSubtype);
         }
-        return ComplexSemType.createComplexSemType(all, subtypes);
     }
 
-    public static SubtypeData createXmlSubtype(boolean isRo, int primitives, Bdd sequence) {
-        int mask = isRo ? XML_PRIMITIVE_RO_MASK : XML_PRIMITIVE_RW_MASK;
-        int p = primitives & mask;
-        if (sequence instanceof BddAllOrNothing && ((BddAllOrNothing) sequence).isAll() && p == mask) {
+    public static SubtypeData createXmlSubtype(int primitives, Bdd sequence) {
+        int p = primitives & XML_PRIMITIVE_ALL_MASK;
+        if (sequence instanceof BddAllOrNothing && ((BddAllOrNothing) sequence).isAll() &&
```

Comment:
I feel like if we use the pattern matching here we can make this more readable pattern match here as well. WDYT?
```suggestion
        if (sequence instanceof BddAllOrNothing bdd && bdd.isAll() &&
```